[WIP] Add purchase_amendment

guewen opened the pull request:
**Work in progress**

Depends on
- [ ] https://github.com/odoo/odoo/pull/6038
- [ ] OCA/OCB#208
- [ ] https://github.com/OCA/stock-logistics-workflow/pull/94

Commit:
Started purchase_amendment on the model of sale_amendment

Commit:
Add test_ship_and_cancel_part

Commit:
Add test test_amend_ship_half and fix a function name

Commit:
Add test_amend_revert_cancel

Commit:
Purchase amendment is no longer reserved to shipping exception

It is not triggered by the cancellation of stock moves, but by a button on the
purchase. So any line, besides received and canceled ones, can be amended.

Commit:
Move the procurements on the new line

Commit:
Keep a link to the amended lines

It allows not to display 2 lines for the same products in the amendment wizard
when part of the products have been shipped and the other one amended.

Commit:
Fix purchase state and creation of moves when fully amending a cancellation

Commit:
Move the amend button so it always appears at the end

before the cancel button

Commit:
Add view of lines grouped by state and order

Commit:
Invoiced purchase cannot be amended

The previous check was based on the 'paid' state of the invoices

Commit:
Usage in the readme

Commit:
Add translation template

pedrobaeza commented on purchase_amendment/README.rst:
```diff
@@ -0,0 +1,52 @@
+Purchase Amendment
+==============
```
same length of title

pedrobaeza commented on purchase_amendment/README.rst:
```diff
@@ -0,0 +1,52 @@
+changes his mind about a purchase.
+
+See also `sale_amendment` (in https://github.com/OCA/sale-workflow)
+which cover a similar use case for changes in sale orders.
```
s/cover/covers

pedrobaeza commented on purchase_amendment/README.rst:
```diff
@@ -0,0 +1,52 @@
+which cover a similar use case for changes in sale orders.
+
+
+Background information
```
This section can be dropped

pedrobaeza commented on purchase_amendment/model/purchase_order_amendment.py:
```diff
@@ -0,0 +1,297 @@
+
+        items = []
+        for line in purchase.order_line:
+            if line.state in ('cancel', 'done'):
```
`if line.state in ('cancel', 'done') or line.amended_by_ids:`

gurneyalex committed:
README fixes

gurneyalex commented:
:+1:

jgrandguillaume committed:
Allow to amend PO with invoices if policy is 'based on incoming shipment'

jgrandguillaume committed:
fix chatter message

gurneyalex committed:
[FIX] unexpected except_picking state

the PO would get in except picking state if amended + shipped
because there would be some canceled stock moves (linked to
canceled po lines).

gurneyalex committed:
[FIX] amending after partial ship

gurneyalex committed:
[ADD] test partial ship + amend + partial ship + amend to 0

gurneyalex committed:
fixup! [ADD] test partial ship + amend + partial ship + amend to 0

gurneyalex committed:
XXX alter stock-logistics-workflow branch until https://github.com/OCA/stock-logistics-workflow/pull/94 is merged

gurneyalex committed:
fixup! fixup! [ADD] test partial ship + amend + partial ship + amend to 0

gurneyalex committed:
WIP purchase order amendment tests

gurneyalex committed:
fixup! WIP purchase order amendment tests

gurneyalex committed:
fixup! fixup! WIP purchase order amendment tests

lepistone committed:
fix split of procurements, enable test

lepistone committed:
fix comment

lepistone committed:
bring in test_cancel_one_line

lepistone committed:
lint

lepistone committed:
new generic record assertion, and new test

At some point we should be able to extract that assertion method, and
get rid of most of the custom helpers.

lepistone commented:
superseded by #138
